### PR TITLE
Amount secret pr

### PR DESCRIPTION
**Pull request by dan-da**
Changes to support sn_payment crate:

1. fix: adds commitment and range_proof fields to DbcContent hash().
2. Add hash fn to DerivedKeySet, DbcPacket.
3. move encrypt, decrypt, gen pedersen commitments into AmountSecret

I think it makes the API a bit cleaner and more "rusty" overall.

### CODE REVIEW

File: src/dbc_content.rs [hunk 1]
```diff
@@ -67,6 +68,14 @@ impl AmountSecrets {
...
     }
 
+    pub fn to_pedersen_commitment(&self) -> RistrettoPoint {
+        PedersenGens::default().commit(Scalar::from(self.amount), self.blinding_factor)
+    }
+
+    pub fn to_pedersen_commitment_compressed(&self) -> CompressedRistretto {
+        self.to_pedersen_commitment().compress()
+    }
+
```
Comment (davidrusu) [inline src/dbc_content.rs:78]: I think just `to_pedersen_commitment()` is enough, if someone wants the Compressed point, they can compress it themselves

File: src/dbc_content.rs [hunk 2]
```diff
@@ -67,6 +68,14 @@ impl AmountSecrets {
         }
     }
 
+    pub fn to_pedersen_commitment(&self) -> RistrettoPoint {
+        PedersenGens::default().commit(Scalar::from(self.amount), self.blinding_factor)
+    }
+
+    pub fn to_pedersen_commitment_compressed(&self) -> CompressedRistretto {
+        self.to_pedersen_commitment().compress()
+    }
```
Comment (davidrusu) [inline src/dbc_content.rs:77]: Same here, we can keep the API small and let users compress if they need to

**Review by davidrusu — DISMISSED** [review]

File: src/dbc_content.rs (hunk 1 above)
Comment (dan-da) [reply src/dbc_content.rs]: done.

**Comment by dan-da** [comment]
@davidrusu This PR was previously introducing a pattern of ::to_bytes() -> Vec<u8> and ::hash() -> Hash for many structs.   sn_payment is also doing a similar thing.    The hash function was always exactly the same, it just calls to_bytes() and hashes it.  This repeating boilerplate code seemed icky, so instead I made a trait `ByteHash` which impls a default hash() fn and requires to_bytes().   I think it is nicer, but let me know what you think.

One thing I don't have a great solution for is from_bytes(), ie for symmetry.   That would require more rigor to account for the exact sizes of every field.  And we don't need from_bytes() for generating a hash or anything else at present.   So I did not make it part of the trait.

For the most part, behavior remains the same.  One thing that changes is that dbc::spend_key_index() was using the hash of DbcContent and DbcTransaction as byte input to the hash.    Now it uses their to_bytes() instead.  I think that should be fine/equivalent, but figured it is worth noting.

**Comment by dan-da** [comment]
obsoleted by ringct.   closing.

**Pull request closed without merging**